Add command-line functionality

emjbishop opened the pull request:
This PR indroduces command-line functionality, allowing users to perform the same tasks as before but through the terminal.

**Changes**
- New CLI functionality:
    - Added `cli.py`: Contains CLI entry point (using `click`).
    - Added `build_lookup.build_lookup_from_fastas_cli()`: Runs the existing `build_lookup.build_lookup_from_fastas()` using terminal inputs.
    - Added `convert.convert_gene_cli()`: Reads an input file into a DataFrame, runs the existing `convert.convert_gene()` using terminal inputs, then saves the output to an output file.

- Documentation updates:
    - Replaced the `usage.ipynb` notebook with two separate notebooks:
        - `usage_cli.ipynb` for CLI usage.
        - `usage_library.ipynb` for library usage.
    - Extensively updated the documentation to give CLI instructions and examples, and to be neutral towards people using TCRconvert via importing or the command-line.

- Dependencies and formatting:
    - Added `click>=8.1.7` as an installation requirement.
    - Added extra blank lines at the ends of some example files to match how Pandas writes output CSV/TSV files.

**Testing**
- Added `test_cli.py`: Contains tests for CLI functionality (using `click.testing`).
- Verified CLI functionality by running commands with various settings.

@sckott if you have time could you try installing from this branch and just ensure that the terminal options show up?

sckott commented:
Will do, thanks!

sckott commented on tcrconvert/convert.py:
```diff
@@ -175,3 +177,55 @@ def convert_gene(df, frm, to, species='human', frm_cols=[], quiet=False):
+    # Save output
+    if outfile.endswith('csv'):
+        out_df.to_csv(outfile, index=False)
+    elif outfile.endswith('tsv'):
```
same with next if else block and throughout package

emjbishop replied:
I tried this out but got a syntax error and so I'm sticking with `elif` (instead of multiple `if`s) for now.

sckott replied:
what code did you try that gave the syntax error? perhaps you still tried to use a condition with `else`? `else would be used like 

```python
if outfile.endswith('csv'):
  out_df.to_csv(outfile, index=False)
else:
  out_df.to_csv(outfile, sep='\t', index=False)
```

that is, you can't pass a condition to `else`. Another approach might be to skip the if else and do:

```python
sep = "," if outfile.endswith("csv") else "\t"
out_df.to_csv(outfile, sep=sep, index=False)
```

just another option, do whatever feels best

emjbishop replied:
Updated in d096361249ec3eb0dbac38899743b6aca0ead37a

sckott commented on tcrconvert/convert.py:
```diff
@@ -175,3 +177,55 @@ def convert_gene(df, frm, to, species='human', frm_cols=[], quiet=False):
+    if outfile.endswith('csv'):
+        out_df.to_csv(outfile, index=False)
+    elif outfile.endswith('tsv'):
+        out_df.to_csv(outfile, sep='\t', index=False)
```
might want to `click.echo()` a message that the file has been written to a the path that's declared by `outfile`

emjbishop replied:
- Added this as part of printed info to CLI users with `click.echo` in 294aeba1c7dac5a25a991449e00c6a90392c7d39

sckott commented on tcrconvert/cli.py:
```diff
@@ -0,0 +1,15 @@
+import click
```
One thing you can do to provide info to users of the cli tool is use `click.echo` (see https://click.palletsprojects.com/en/stable/quickstart/#echoing) and you can even style echo output with colors, etc. if you want to

emjbishop replied:
- Added link to click's autocompletion documentation to the docs in 1d34d365f82a0aff40b078d3cdffff0bfa10d9dc 
- Added printed info to CLI users with `click.echo` in 294aeba1c7dac5a25a991449e00c6a90392c7d39

sckott commented:
- Python version
  - I would encourage you to use a more recent Python if possible. 3.10 is getting pretty old. We're on 3.13 as the current one now. 
  - I'd strongly suggest pinning the Python version in your package with a `.python-version` file
- Dependencies
  - I couldn't run tests from this PR, so I had to change a few things
  - You can do away with the requirements.txt files and just list dependencies in your pyproject.toml with sections for the main code to run, for dev and for docs 
  - I'd pin each dependency to a min version, e.g., `click >= 8.1.7`
  - You're missing `click` in `dependencies`  - I see you have it in `requirements.txt` but you don't have it in your pyproject.toml. 
  - I'd make sure to use dev dependencies. If you switch to using `uv` then you can add dev dependencies by doing e.g., `uv add pytest --dev`
- Docs and examples for functions
  - I'd strongly recommend having examples fully work. E.g., you have `pd.read_csv('path/to/10x_tcrs.csv')` - instead have an example file that comes with the package that the user can use, If you have code that can really run you could optionally test that your excamples run too with https://docs.python.org/3/library/doctest.html 
- Code in general
  - strongly recommend avoiding single letter parameters in your classes/functions, e.g., `pad_single_digit` and `add_dash_one`

sckott reviewed (commented):
added some comments on certain lines and in general

emjbishop commented on tcrconvert/convert.py:
```diff
@@ -175,3 +177,55 @@ def convert_gene(df, frm, to, species='human', frm_cols=[], quiet=False):
+
+
+# Command-line version of convert_gene()
+@click.command(name='convert-gene', no_args_is_help=True)
```
Shortened to `convert` and `build` in 323e90d7b411f16d1f839a6694758fb82640b507

emjbishop commented on tcrconvert/convert.py:
```diff
@@ -175,3 +177,55 @@ def convert_gene(df, frm, to, species='human', frm_cols=[], quiet=False):
+
+    # Check that input and output paths are CSV/TSV
+    if not infile.endswith(('csv', 'tsv')):
+        logger.error('"infile" must be a .csv or .tsv file')
```
Switched to using `click` error handling for input paths in 7dad2a2e427e483b065786e86ce1f304af15d797

emjbishop commented on tcrconvert/convert.py:
```diff
@@ -175,3 +177,55 @@ def convert_gene(df, frm, to, species='human', frm_cols=[], quiet=False):
+def convert_gene_cli(infile, outfile, frm, to, species, frm_cols, quiet):
+    '''Convert T-cell receptor V/D/J/C gene names.
+
+    :Example:
```
Added example files in 5dec475a8bf20299303ffe54abbf1121470f809a and ensured docstrings work as examples in f8d6a5022499af49f23062333dee6dd35d04f2bd (among other commits to ensure working examples).

emjbishop commented on tcrconvert/convert.py:
```diff
@@ -175,3 +177,55 @@ def convert_gene(df, frm, to, species='human', frm_cols=[], quiet=False):
+
+    .. code-block:: bash
+
+       $ tcrconvert convert-gene --infile 10x_tcrs.csv --outfile converted.tsv --frm tenx --to adaptive --species mouse -c myV -c myD -c myJ --quiet
```
Added example files in 5dec475a8bf20299303ffe54abbf1121470f809a and ensured docstrings work as examples in f8d6a5022499af49f23062333dee6dd35d04f2bd (among other commits to ensure working examples).

emjbishop commented:
Regarding the 11/21/24 [comment](https://github.com/seshadrilab/tcrconvert/pull/3#issuecomment-2492626623):

- Python version
  - Fixed by pinning version to 3.12 in 612254ffacb958e933f468005a82f480397598c3
- Dependencies
  - Fixed via multiple commits, now all requirements are specified in the pyproject.toml
- Docs and examples for functions
  - Fixed via multiple commits, now we have working examples
- Code in general
  - Fixed single letter parameters in c238ce605d36e2030d5433f783a00dfb9d8d37d4
